app/experimenter/nimbus-ui/src/components/FormBranches/index.tsx

```diff
@@ -6,43 +6,114 @@ import React from "react";
   experiment,
-  equalRatio,
   featureConfig,
+  onSave,
+  onNext,
 }: {
   experiment: getExperiment_experimentBySlug;
-  equalRatio?: boolean;
   featureConfig: getConfig_nimbusConfig["featureConfig"];
+  onSave: (state: FormBranchesSaveState) => void;
+  onNext: () => void;
 }) => {
+  const [formBranchesState, dispatch] = useFormBranchesReducer(experiment);
   const {
+    featureConfig: experimentFeatureConfig,
     referenceBranch,
     treatmentBranches,
-    featureConfig: experimentFeatureConfig,
-  } = experiment!;
+    equalRatio,
+  } = formBranchesState;
+
+  const handleAddBranch = () => dispatch({ type: "addBranch" });
+
+  const handleRemoveBranch = (idx: number) => () =>
+    dispatch({ type: "removeBranch", idx });
+
+  const handleUpdateBranch = (idx: number) => (value: AnnotatedBranch) =>
+    dispatch({ type: "updateBranch", idx, value });
+
+  const handleEqualRatioChange = (ev: React.ChangeEvent<HTMLInputElement>) =>
+    dispatch({ type: "setEqualRatio", value: ev.target.checked });
+
+  // HACK: just use the first available feature config when adding
+  // The only available indicator whether to display "Add feature config" is a non-null feature config
+  const handleAddFeatureConfig = () =>
+    dispatch({ type: "setFeatureConfig", value: featureConfig![0] });
+
+  const handleRemoveFeatureConfig = () =>
+    dispatch({ type: "removeFeatureConfig" });
+
+  const handleFeatureConfigChange = (
+    value: getConfig_nimbusConfig_featureConfig | null,
+  ) => dispatch({ type: "setFeatureConfig", value });
+
+  // TODO: EXP-505 implement save button enable/disable logic
+  const isSaveDisabled = false;
+
+  const handleSaveClick = (
+    ev: React.MouseEvent<HTMLButtonElement, MouseEvent>,
+  ) => {
+    ev.preventDefault();
+    onSave(extractSaveState(formBranchesState));
+  };
+
+  // TODO: EXP-505 implement next button enable/disable logic
+  const isNextDisabled = false;
+
+  const handleNextClick = (
+    ev: React.MouseEvent<HTMLButtonElement, MouseEvent>,
+  ) => {
+    ev.preventDefault();
+    onNext();
+  };
+
+  const commonBranchProps = {
+    equalRatio,
+    featureConfig,
+    experimentFeatureConfig,
+    onFeatureConfigChange: handleFeatureConfigChange,
+    onAddFeatureConfig: handleAddFeatureConfig,
+    onRemoveFeatureConfig: handleRemoveFeatureConfig,
+  };
+
   return (
     <section data-testid="FormBranches" className="border-top my-3">
       <Form className="p-2">
         <Form.Row className="my-3">
           <Form.Group controlId="evenRatio">
             <Form.Check
+              data-testid="equal-ratio-checkbox"
+              onChange={handleEqualRatioChange}
+              checked={equalRatio}
               type="checkbox"
               label="Users should be split evenly between all branches"
             />
           </Form.Group>
           <Form.Group as={Col} className="align-top text-right">
-            <Button variant="outline-primary" size="sm">
+            <Button
```

Comment:
@lmorchard It looks like, if you start with control+treatment branch, and then either 1) remove the treatment and click add branch, or 2) just click add branch twice, it adds duplicate branches 🤔

Reply:
I'm not seeing this happening - how are you reproducing it exactly?

Reply:
Cannot reproduce this in Storybook. Did a full re-install of Experimenter and can finally see this happening when served up from the full app context with data from Django.

Not sure exactly what the problem is yet, but it seems like somehow the reducer `dispatch` function is getting called twice for adding a branch. But, only from the real server, not Storybook.

Reply:
Looks like I've hit [this behavior in `useReducer`](https://github.com/facebook/react/issues/16295) and have some unintended mutations in the reducer.